### PR TITLE
[fix](kerberos)fix and refactor ugi login for kerberos and simple authentication

### PR DESCRIPTION
**Pull request by wsjz**
## Proposed changes
optimize kerberos ugi login:
1. support authentication framework for external table
2. cache ugi to avoid conflicts when set configuration for ugi login 
3. do ugi login just on creating catalog, which reduces the ugi create times 
4. only simple authentication will use getloginUser, which avoids conflicts between simple and kerberos authentication

### CODE REVIEW

**Comment by wsjz** [comment]
run buildall

**Comment by wsjz** [comment]
run buildall

**Comment by wsjz** [comment]
run buildall

**Comment by wsjz** [comment]
run buildall

**Comment by wsjz** [comment]
run buildall

**Comment by wsjz** [comment]
run external

**Comment by wsjz** [comment]
run buildall

File: fe/fe-common/src/main/java/org/apache/doris/common/security/authentication/HadoopKerberosAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            nextRefreshTime = calculateNextRefreshTime(newSubject);
+        }
+        return ugi;
+    }
+
+    private UserGroupInformation login(Subject subject) throws IOException {
+        // login and get ugi when catalog is initialized
+        initializeAuthConfig(config.getConf());
+        String principal = config.getKerberosPrincipal();
+        LOG.debug("Login by kerberos authentication with principal: {}", principal);
```
Comment (morningman) [inline fe/fe-common/src/main/java/org/apache/doris/common/security/authentication/HadoopKerberosAuthenticator.java:95]: `if (LOG.isDebugEnable())`

File: fe/fe-common/src/main/java/org/apache/doris/common/security/authentication/HadoopSimpleAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+public class HadoopSimpleAuthenticator implements HadoopAuthenticator {
+    private static final Logger LOG = LogManager.getLogger(HadoopSimpleAuthenticator.class);
+    private final UserGroupInformation ugi;
+
+    public HadoopSimpleAuthenticator(SimpleAuthenticationConfig config) {
+        String hadoopUserName = config.getUsername();
+        if (hadoopUserName == null) {
+            hadoopUserName = "hadoop";
+            config.setUsername(hadoopUserName);
+            LOG.debug("{} is unset, use default user: hadoop", AuthenticationConfig.HADOOP_USER_NAME);
```
Comment (morningman) [inline fe/fe-common/src/main/java/org/apache/doris/common/security/authentication/HadoopSimpleAuthenticator.java:35]: All debug log should be wrapped with `if (LOG.isDebugEnable())`

File: fe/fe-common/src/main/java/org/apache/doris/common/security/authentication/HadoopSimpleAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        LOG.debug("Login by proxy user, hadoop.username: {}", hadoopUserName);
+    }
+
+    @Override
+    public UserGroupInformation getUGI() {
+        return ugi;
+    }
+
+    @Override
+    public <T> T doAs(PrivilegedExceptionAction<T> action) throws Exception {
```
Comment (morningman) [inline fe/fe-common/src/main/java/org/apache/doris/common/security/authentication/HadoopSimpleAuthenticator.java:47]: why not using parent method?

File: fe/fe-common/src/main/java/org/apache/doris/common/security/authentication/ImpersonatingHadoopAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    @Override
+    public UserGroupInformation getUGI() throws IOException {
+        synchronized (ImpersonatingHadoopAuthenticator.class) {
+            ugi = UserGroupInformation.createProxyUser(username, delegate.getUGI());
+        }
+        return ugi;
+    }
+
+    @Override
+    public <T> T doAs(PrivilegedExceptionAction<T> action) throws Exception {
```
Comment (morningman) [inline fe/fe-common/src/main/java/org/apache/doris/common/security/authentication/ImpersonatingHadoopAuthenticator.java:46]: use parent method

File: fe/fe-common/src/main/java/org/apache/doris/common/security/authentication/ImpersonatingHadoopAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    public ImpersonatingHadoopAuthenticator(HadoopAuthenticator delegate, String username) {
+        this.delegate = Objects.requireNonNull(delegate);
+        this.username = Objects.requireNonNull(username);
+    }
+
+    @Override
+    public UserGroupInformation getUGI() throws IOException {
+        synchronized (ImpersonatingHadoopAuthenticator.class) {
+            ugi = UserGroupInformation.createProxyUser(username, delegate.getUGI());
```
Comment (morningman) [inline fe/fe-common/src/main/java/org/apache/doris/common/security/authentication/ImpersonatingHadoopAuthenticator.java:40]: create ugi for each call?

File: fe/fe-core/src/main/java/org/apache/doris/datasource/hive/HiveMetaStoreCache.java [hunk 6]
```diff
@@ -163,6 +167,7 @@ public Map<PartitionCacheKey, HivePartition> loadAll(Iterable<? extends Partitio
         }, null, refreshExecutor);
 
         setNewFileCache();
+        initAuthenticator();
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/hive/HiveMetaStoreCache.java:170]: Move this to ExternalCatalog, this is nothing to do with `cache`

File: fe/fe-core/src/main/java/org/apache/doris/datasource/hive/HiveMetaStoreCache.java [hunk 7]
```diff
@@ -1071,6 +1085,44 @@ private static boolean isGeneratedPath(String name) {
         }
     }
 
+    @Data
+    public static class AuthenticatorCacheKey {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/hive/HiveMetaStoreCache.java:1089]: This class can be removed

**Comment by wsjz** [comment]
run buildall

**Comment by wsjz** [comment]
run buildall

**Comment by wsjz** [comment]
run buildall

**Comment by wsjz** [comment]
run buildall

**Comment by wsjz** [comment]
run buildall

**Comment by wsjz** [comment]
run external

**Comment by wsjz** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Review by hello-stephen — APPROVED** [review]

**Review by wuwenchi — APPROVED** [review]

**Pull request merged**